cpp/tests/test_abm_serialization.cpp

```diff
@@ -0,0 +1,298 @@
+ * i.e. the function does not contain conditionals (`if (t > 0)`), optionals (`add_optional`/`expect_optional`), etc.,
+ * and assuming that deserialize is injective (meaning that two unequal instances of T do not have the same json
+ * representation, which can happen e.g. if not all member variables are serialized),
+ * this sufficiently tests that serialize and deserialize are inverse functions to each other.
+ *
+ * @tparam T The type to test.
+ * @param reference_json A json value representing an instance of T.
+ */
+template <class T>
+void test_json_serialization_by_representation(const Json::Value& reference_json)
+{
+    // check that the json is deserializable (i.e. a valid representation)
+    auto t_result = mio::deserialize_json(reference_json, mio::Tag<T>());
+    ASSERT_THAT(print_wrap(t_result), IsSuccess());
+
+    // check that the resulting type T is serializable
+    auto json_result = mio::serialize_json(t_result.value());
+    ASSERT_THAT(print_wrap(json_result), IsSuccess());
+
+    test_equal_json_representation(json_result.value(), reference_json);
+}
+
+/**
+ * @brief Test de- and serialization of an object by comparing its json representation and using its equality operator.
+ *
+ * First, test that serializing the reference_object is equal to the reference_json (w.r.t. their representation),
+ * and that deserializing the reference_json results in an object equal to the reference_object.
+ * Then, repeat this step using its own results as arguments to (de)serialize, to check that serialization and
+ * deserialization are inverse functions to each other.
+ *
+ * @tparam T The type to test.
+ * @param reference_object An instance of T.
+ * @param reference_json A json value representing reference_object.
+ */
+template <class T>
+void test_json_serialization_full(const T& reference_object, const Json::Value& reference_json)
+{
+    // check that the reference type T is serializable
+    auto json_result = mio::serialize_json(reference_object);
+    ASSERT_THAT(print_wrap(json_result), IsSuccess());
+
+    // check that the reference json is deserializable
+    auto t_result = mio::deserialize_json(reference_json, mio::Tag<T>());
+    ASSERT_THAT(print_wrap(t_result), IsSuccess());
+
+    // compare both results with other reference values
+    EXPECT_EQ(t_result.value(), reference_object);
+    test_equal_json_representation(json_result.value(), reference_json);
+
+    // do the same once more using the results from above
+    auto json_result_2 = mio::serialize_json(t_result.value());
+    ASSERT_THAT(print_wrap(json_result_2), IsSuccess());
+    auto t_result_2 = mio::deserialize_json(json_result.value(), mio::Tag<T>());
+    ASSERT_THAT(print_wrap(t_result_2), IsSuccess());
+
+    EXPECT_EQ(t_result_2.value(), reference_object);
+    test_equal_json_representation(json_result_2.value(), reference_json);
+}
```

Comment:
You are right - I will remove the second half.

Reply:
Actually, when removing the second half, is there a benefit to testing `j == serialize(t) && t == deserialize(j)` over just `j == serialize(deserialize(j))`?
The second half is useful (or needed) if T::operator== does not cover the same members as serialization, like Location does (equality only checks the id). But we cannot test for that, I think.
So should I keep the second half, remove this test entirely, or just document the requirement on "=="?

Reply:
I had to think about this for a while, it's very confusing...

I think that `j == serialize(deserialize(j))` works. `t == deserialize(serialize(t))` would not work, since it doesn't check the data types of serialized members. 

 `j == serialize(deserialize(j))` also doesn't require the serialized type to have operator==, so works for every type. this is another advantage.

If a type has operator==, then `t == deserialize(serialize(t))` should always be true. not sure whether to test this here, since it's really a test of operator== and not of serialization, but maybe it adds a bit of confidence? Anyway, `j == serialize(t) && t == deserialize(j)` should be sufficient for both `t == deserialize(serialize(t))` and  `j == serialize(deserialize(j))`.

Reply:
to summarize, `j == serialize(deserialize(j))` should be enough.